### PR TITLE
Group traversal simplify upsert

### PR DESCRIPTION
**Pull request by jon-betts**
Previously:

* The `GroupRoute` looked up groups and returned them (like a service)
* It failed if the group was none with `KeyError`
* The `GroupUpsertRoute` used the `GroupRoute` and then created a `GroupUpsertContext` for it
* The `GroupUpsertRoute` had to catch the error and use `None` for the group instead

Now:

* The `GroupUpsertRoute` is called `GroupRoute` and is responsible for looking up the group
* The `GroupUpsertContext` is now called `GroupContext` and the group is optional (and it's a data class)
* The `GroupRoute` is now called `GroupRequiredRoute` and is a child of `GroupRoute`
* It's only job is to check if the group context has a missing group and fail

### Why bother?

* Most of the simplification here came from cutting down on test and comment waffle
* But... this is moving towards a world where the context is always required
* We don't want one context for groups and another for group upserts

### CODE REVIEW

File: h/traversal/__init__.py [hunk 1]
```diff
@@ -75,9 +75,8 @@
     "AnnotationRoot",
     "AuthClientRoot",
     "BulkAPIRoot",
+    "GroupRequiredRoot",
     "GroupRoot",
-    "GroupUpsertContext",
-    "GroupUpsertRoot",
```
Comment (jon-betts) [inline h/traversal/__init__.py:80]: Nobody should really be using the context directly outside of tests

File: tests/h/traversal/organization_test.py [hunk 2]
```diff
@@ -24,10 +24,10 @@ def test_it_404s_if_the_organization_doesnt_exist(
         with pytest.raises(KeyError):
             OrganizationRoot(pyramid_request)[sentinel.non_existent_pubid]
 
-    @pytest.fixture
+    @pytest.fixture(autouse=True)
     def with_noise_organization(self, factories):
         # Add a handful of organizations to the DB to make the test realistic.
-        factories.Organization.generate_batch(size=2)
+        factories.Organization.create_batch(size=2)
```
Comment (jon-betts) [inline tests/h/traversal/organization_test.py:30]: This is a coverage fix, but actually this was a genuine mistake introduced in the previous PR. The fixture wasn't used and didn't work if you called it... great job... slow clap me.

File: tests/h/presenters/annotation_base_test.py [hunk 3]
```diff
@@ -111,12 +111,6 @@ def test_target_missing_selectors(self):
 class Berlin(datetime.tzinfo):
     """Berlin timezone, without DST support"""
 
-    def utcoffset(self, dt):
-        return datetime.timedelta(hours=1)
-
-    def tzname(self, dt):
-        return "Berlin"
-
```
Comment (jon-betts) [inline tests/h/presenters/annotation_base_test.py:119]: Small unrelated coverage fix to keep our stats up

File: h/traversal/group.py [hunk 4]
```diff
@@ -1,86 +1,55 @@
...
-        The "upsert" permission is an unusual hybrid. It has a different meaning
-        depending on the upsert situation.
-
-        If there is no group associated with the context, the "upsert" permission
-        should be given to all real users such that they may use the UPSERT endpoint
-        to create a new group. However, if there is a group associated with the
-        context, the "upsert" permission is managed by the model. The model only
-        applies "upsert" for the group's creator. This will allow the endpoint to
-        support updating a specific group (model), but only if the request's
-        user should be able to update the group.
```
Comment (jon-betts) [inline h/traversal/group.py:78]: I don't think it's beneficial to explain the contents of the ACLs in another file partially because it's long, and partially because it's easy for this to end up being a lie. The rest I think has been summarised below in fewer words.

File: tests/h/routes_test.py [hunk 5]
```diff
@@ -132,19 +132,19 @@ def test_includeme():
             request_method="POST",
             factory="h.traversal.BulkAPIRoot",
         ),
-        call("api.groups", "/api/groups", factory="h.traversal.GroupRoot"),
+        call("api.groups", "/api/groups", factory="h.traversal.GroupRequiredRoot"),
```
Comment (jon-betts) [inline tests/h/routes_test.py:135]: This never uses the subscript look-up, so actually either would do, but a group is certainly not required here.

**Review by seanh — APPROVED** [review]
